extractor/image/image.go

```diff
@@ -0,0 +1,240 @@
+package image
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+
+	"github.com/containers/image/image"
+	"github.com/containers/image/pkg/blobinfocache"
+	"github.com/containers/image/pkg/compression"
+	"github.com/containers/image/transports/alltransports"
+	imageTypes "github.com/containers/image/types"
+	"github.com/docker/distribution/reference"
+	"github.com/opencontainers/go-digest"
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/fanal/cache"
+	"github.com/aquasecurity/fanal/types"
+)
+
+type ImageSource interface {
+	GetBlob(ctx context.Context, info imageTypes.BlobInfo, cache imageTypes.BlobInfoCache) (reader io.ReadCloser, n int64, err error)
+}
+
+type ImageCloser interface {
+	LayerInfos() (layerInfos []imageTypes.BlobInfo)
+	ConfigBlob(ctx context.Context) (blob []byte, err error)
+}
+
+type Reference struct {
+	Name   string
+	IsFile bool
+}
+
+type Image struct {
+	name       string   // e.g. alpine:3.10
+	isFile     bool     // from a tar file
+	transports []string // e.g. "docker://"
+
+	systemContext *imageTypes.SystemContext
+	blobInfoCache imageTypes.BlobInfoCache
```

Comment:
Question: What are they used for? It's not clear to me reading it right away.

Reply:
`imageTypes.BlobInfoCache` can be used to store a relationship between compressed and decompressed layer IDs. We sometimes need to know a decompressed layer ID for cache without decompressing a layer actually. I introduced it once, but I removed the implementation to keep this PR small. I will add it again in the next PR. If this bothers you, I can remove it for now.

Reply:
Let's add this as part of a separate PR. Would be easier to keep track that way.

Reply:
Sorry, we need it here.
https://github.com/aquasecurity/fanal/pull/70/files#diff-525dd35201b72b43ad63b85dd2720ba6R224